File: SceytChatUiKit/src/main/java/com/sceyt/sceytchatuikit/media/audio/AudioRecorderHelper.java
package com.sceyt.sceytchatuikit.media.audio;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.File;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class AudioRecorderHelper {

    private final static Executor recorderExecutor = Executors.newSingleThreadScheduledExecutor();
    private static File audioFile;
    private static AudioRecorder currentRecorder;
    private static final Gson serializer = new GsonBuilder().create();

    public interface OnRecorderStart {
        void onStart(boolean started);
    }

    public interface OnRecorderStop {
        void onStop(boolean tooShort, File recordedFile, Integer duration, Integer[] amplitudes);
    }

    public interface OnRecorderCancel {
        void onCancel();
    }

    public static void startRecording(String directoryToSaveFile, OnRecorderStart onRecorderStart) {
        recorderExecutor.execute(() -> {
            audioFile = FileManager.createFile(AudioRecorderImpl.AUDIO_FORMAT, directoryToSaveFile);
            currentRecorder = new AudioRecorderImpl(audioFile);
            boolean started = currentRecorder.startRecording(32000, null);
            if (onRecorderStart != null) {
                onRecorderStart.onStart(started);
            }
        });
    }

    public static void stopRecording(OnRecorderStop onRecorderStop) {
        recorderExecutor.execute(() -> {
            currentRecorder.stopRecording();

            int duration = getCurrentDuration();
            boolean isTooShort = false;
            if (duration < 1) {
                audioFile.delete();
                isTooShort = true;
            }

            if (onRecorderStop != null) {
                onRecorderStop.onStop(isTooShort, audioFile, duration, getCurrentAmplitudes());
            }
        });
    }

    public static void cancelRecording(OnRecorderCancel onRecorderCancel) {
        recorderExecutor.execute(() -> {
            currentRecorder.stopRecording();
            audioFile.delete();
            if (onRecorderCancel != null)
                onRecorderCancel.onCancel();
        });
    }

    public static Integer[] getCurrentAmplitudes() {
        if (currentRecorder != null)
            return currentRecorder.getRecordingAmplitudes();
        else
            return new Integer[]{0};
    }

    public static Integer getCurrentDuration() {
        if (currentRecorder != null)
            return currentRecorder.getRecordingDuration();
        else
            return 0;
    }

    public static String getJsonAmplitudes() {
        Integer[] amps = getCurrentAmplitudes();
        return serializer.toJson(amps, Integer[].class);
    }

}

File: SceytChatUiKit/src/main/java/com/sceyt/sceytchatuikit/media/audio/AudioRecorderHelper.kt
package com.sceyt.sceytchatuikit.media.audio

import android.content.Context
import com.google.gson.GsonBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.concurrent.Executors

class AudioRecorderHelper(
        private val scope: CoroutineScope,
        private val context: Context
) {
    private val recorderDispatcher = Executors.newSingleThreadScheduledExecutor().asCoroutineDispatcher()
    private var audioFile: File? = null
    private var currentRecorder: AudioRecorder? = null
    private val serializer = GsonBuilder().create()
    private val audioFocusHelper: AudioFocusHelper by lazy { AudioFocusHelper(context) }

    fun startRecording(directoryToSaveFile: String, onRecorderStart: OnRecorderStart? = null) {
        scope.launch(recorderDispatcher) {
            audioFocusHelper.requestAudioFocusCompat()
            val audioFile = FileManager.createFile(AudioRecorderImpl.AUDIO_FORMAT, directoryToSaveFile).also {
                this@AudioRecorderHelper.audioFile = it
            }
            val recorder = AudioRecorderImpl(context, audioFile).also { currentRecorder = it }
            val started = recorder.startRecording(32000, null)
            withContext(Dispatchers.Main) {
                onRecorderStart?.onStart(started)
            }
        }
    }

    fun stopRecording(onRecorderStop: OnRecorderStop? = null) {
        scope.launch(recorderDispatcher) {
            currentRecorder?.stopRecording()
            val duration = currentDuration
            var isTooShort = false
            if (duration < 1) {
                audioFile?.delete()
                isTooShort = true
            }
            audioFocusHelper.abandonCallAudioFocusCompat()
            withContext(Dispatchers.Main) {
                onRecorderStop?.onStop(isTooShort, audioFile, duration, currentAmplitudes)
            }
        }
    }

    fun cancelRecording(onRecorderCancel: OnRecorderCancel? = null) {
        scope.launch(recorderDispatcher) {
            currentRecorder?.stopRecording()
            audioFile?.delete()
            audioFocusHelper.abandonCallAudioFocusCompat()
            withContext(Dispatchers.Main) {
                onRecorderCancel?.onCancel()
            }
        }
    }

    val currentAmplitudes: Array<Int>
        get() = currentRecorder?.getRecordingAmplitudes() ?: arrayOf(0)

    val currentDuration: Int
        get() = currentRecorder?.getRecordingDuration() ?: 0

    val jsonAmplitudes: String
        get() {
            val amps = currentAmplitudes
            return serializer.toJson(amps, Array<Int>::class.java)
        }

    fun interface OnRecorderStart {
        fun onStart(started: Boolean)
    }

    fun interface OnRecorderStop {
        fun onStop(tooShort: Boolean, recordedFile: File?, duration: Int, amplitudes: Array<Int>)
    }

    fun interface OnRecorderCancel {
        fun onCancel()
    }
}